src/traces/funnel/cross_trace_calc.js

```diff
@@ -0,0 +1,74 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var setGroupPositions = require('../bar/cross_trace_calc').setGroupPositions;
+
+module.exports = function crossTraceCalc(gd, plotinfo) {
+    var fullLayout = gd._fullLayout;
+    var fullData = gd._fullData;
+    var calcdata = gd.calcdata;
+    var xa = plotinfo.xaxis;
+    var ya = plotinfo.yaxis;
+    var funnels = [];
+    var funnelsVert = [];
+    var funnelsHorz = [];
+    var cd, i;
+
+    for(i = 0; i < fullData.length; i++) {
+        var fullTrace = fullData[i];
+        var isHorizontal = (fullTrace.orientation === 'h');
+
+        if(
+            fullTrace.visible === true &&
+            fullTrace.xaxis === xa._id &&
+            fullTrace.yaxis === ya._id &&
+            fullTrace.type === 'funnel'
+        ) {
+            cd = calcdata[i];
+
+            if(isHorizontal) {
+                funnelsHorz.push(cd);
+            } else {
+                funnelsVert.push(cd);
+            }
+
+            funnels.push(cd);
+        }
+    }
+
+    // funnel version of 'barmode', 'barnorm', 'bargap' and 'bargroupgap'
+    var mockGd = {
+        _fullLayout: {
+            _axisMatchGroups: fullLayout._axisMatchGroups,
+            _alignmentOpts: fullLayout._alignmentOpts,
+            barmode: fullLayout.funnelmode,
+            barnorm: fullLayout.funnelnorm,
+            bargap: fullLayout.funnelgap,
+            bargroupgap: fullLayout.funnelgroupgap
+        }
+    };
+
+    setGroupPositions(mockGd, xa, ya, funnelsVert);
```

Comment:
Next non-blocking clean-up opportunity: instead of creating a "mock" `gd` object here and for `waterfall` and `barpolar`, we should make `setGroupPosition` accept an `opts` option container with keys `mode`, `norm`, `gap` and `groupgap`. 

Again, this is non-blocking.

Reply:
Good call. May addressed in 75a5cb43c5fca0d1e23a7d7a6c5f5ac053114e4c.

Reply:
I'm not a fan of https://github.com/plotly/plotly.js/commit/75a5cb43c5fca0d1e23a7d7a6c5f5ac053114e4c - we could do better.

This diff in `barpolar/calc.js` is particularly not great:

```js
-   setGroupPositions(mockGd, aAxis, rAxis, barPolarCd);
+   setGroupPositions('bar', mockGd, aAxis, rAxis, barPolarCd);
```

I would much prefer having:

```js
// barpolar/calc.js

setGroupPositions(gd, aAxis, rAxis, {
  mode: polarLayout.barmode,
  norm: polarLayout.barnorm,
  gap: polarLayout.bargap,
  groupgap: polarLayout.bargroupgap
})
```

Reply:
@etpinard Now it should be addressed in https://github.com/plotly/plotly.js/pull/3817/commits/86c0c092c7962d7b66e2131300bb2665b3e39378.
Thanks.

Reply:
https://github.com/plotly/plotly.js/commit/86c0c092c7962d7b66e2131300bb2665b3e39378 is lovely.

Resolved.